executor/src/witgen/jit/affine_symbolic_expression.rs

```diff
@@ -0,0 +1,544 @@
+        AffineSymbolicExpression {
+            coefficients: Default::default(),
+            offset: k,
+        }
+    }
+}
+
+impl<T: FieldElement, V: Ord + Clone + Display> AffineSymbolicExpression<T, V> {
+    pub fn from_known_variable(var: V) -> Self {
+        SymbolicExpression::from_var(var).into()
+    }
+    pub fn from_unknown_variable(var: V) -> Self {
+        AffineSymbolicExpression {
+            coefficients: [(var, T::from(1).into())].into_iter().collect(),
+            offset: SymbolicExpression::from(T::from(0)),
+        }
+    }
+    pub fn from_number(n: T) -> Self {
+        SymbolicExpression::from(n).into()
+    }
+
+    /// If this expression does not contain unknown variables, returns the symbolic expression.
+    pub fn try_to_known(&self) -> Option<&SymbolicExpression<T, V>> {
+        if self.coefficients.is_empty() {
+            Some(&self.offset)
+        } else {
+            None
+        }
+    }
+
+    /// Tries to multiply this expression with another one.
+    /// Returns `None` if the result would be quadratic.
+    pub fn try_mul(&self, other: &Self) -> Option<Self> {
+        if !self.coefficients.is_empty() && !other.coefficients.is_empty() {
+            return None;
+        }
+        let (multiplier, coefficients, offset) = if self.coefficients.is_empty() {
+            (&self.offset, &other.coefficients, &other.offset)
+        } else {
+            (&other.offset, &self.coefficients, &self.offset)
+        };
+        let coefficients = coefficients
+            .iter()
+            .map(|(var, coeff)| (var.clone(), coeff * multiplier))
+            .collect();
+        let offset = offset * multiplier;
+        Some(AffineSymbolicExpression {
+            coefficients,
+            offset,
+        })
+    }
+
+    /// Solves the equation `self = 0` and returns how to compute the solution.
+    pub fn solve(
+        &self,
+        range_constraints: &impl RangeConstraintSet<V, T>,
+    ) -> Result<Vec<Effect<T, V>>, EvalError<T>> {
+        match self.coefficients.len() {
+            0 => {
+                return if self.offset.is_known_zero() {
+                    Ok(vec![])
+                } else {
+                    Err(EvalError::ConstraintUnsatisfiable(self.to_string()))
+                };
+            }
+            1 => {
+                let (var, coeff) = self.coefficients.iter().next().unwrap();
+                assert!(!coeff.is_known_zero());
+                let value = self.offset.field_div(&coeff.neg());
+                let assignment = Effect::Assignment(var.clone(), value);
+                if let Some(coeff) = coeff.try_to_number() {
+                    assert!(!coeff.is_zero(), "Zero coefficient has not been removed.");
+                    return Ok(vec![assignment]);
+                } else {
+                    return Ok(vec![
+                        Assertion::assert_is_nonzero(coeff.clone()),
+                        assignment,
+                    ]);
```

Comment:
I've been thinking about this example:
`instr_add * (a + b - c) + instr_sub * (a - b - c) = 0`

What if both `instr_add` and `instr_sub` are zero?

Assuming the instruction flags and `a` and `b` are known, the affine expression would be:
`(-instr_add - instr_add) * c + (instr_add * (a + b) + instr_sub * (a - b)) = 0`

Solving for `c` yields:
`c := -(instr_add * (a + b) + instr_sub * (a - b)) / (-instr_add - instr_add)`

Which would be a division by zero, which triggers the assertion. I think what this means is that `0 / 0` should be OK, because it means that the constraint has effectively been disabled and the list of effects should be empty!

Reply:
> the list of effects should be empty

I guess not, because we don't know the value of the offset at compile time. Instead, we'd want to generate code like this, right?

```rs
let numerator = -(instr_add * (a + b) + instr_sub * (a - b));
let denominator = (-instr_add - instr_add);
if denominator != 0 {
  c = numerator / denominator;
} else {
  assert_eq!(numerator, 0);
}
```

Reply:
Yeah, just realized the same in a different PR. I would suggest to leave that case open until we hit it, because we actually cannot set `c` to `known` unconditionally.